onecodex/viz/_custom_plots/collection.py

```diff
@@ -0,0 +1,464 @@
+
+        self._classification_fetch()
+        self.rank = self._default_rank
+        self._functional_profiles  # precompute the cache
+
+    def _collate_metadata(self):
+        import pandas as pd
+
+        metadata = [sample._metadata for sample in self.samples]
+
+        if metadata:
+            df = pd.DataFrame(metadata)
+            index = "classification_id" if df["classification_id"].is_unique else "sample_id"
+            metadata = df.set_index(index)
+        else:
+            metadata = pd.DataFrame(
+                columns=["classification_id", "sample_id", "metadata_id", "created_at"]
+            )
+
+        self._cached["metadata"] = metadata
+
+    def _classification_fetch(self):
+        classifications = []
+        job_names = set()
+        for sample in self.samples:
+            summary = sample._classification
+            if not summary:
+                continue
+
+            results = ClassificationResult()
+            results.update(summary["api_results"])
+            results["id"] = summary["uuid"]
+            results.id = summary["uuid"]
+
+            job_names.add(summary["job_name"])
+            classifications.append(results)
+
+        self._cached["is_metagenomic"] = False
+        if all(("One Codex Database" in name for name in job_names)):
+            self._cached["is_metagenomic"] = True
+
+        self._default_rank = self._get_auto_rank(Rank.Auto)
+        self._cached["classifications"] = classifications
+
+    @property
+    def _functional_profiles(self):
+        if "functional_profiles" not in self._cached:
+            functional_results = []
+            for sample in self.samples:
+                profile = sample._functional_profile
+                if not profile:
+                    continue
+                functional_results.append(FunctionalProfiles.from_dict(profile))
+
+            self._cached["functional_profiles"] = functional_results
+
+        return self._cached["functional_profiles"]
+
+    def plot(self, params: PlotParams) -> PlotResult:
+        import altair as alt
+
+        result = None
+        with warnings.catch_warnings(record=True) as captured_warnings:
+            warnings.simplefilter("always", PlottingWarning)
+
+            try:
+                result = self._plot(params)
+            except (ValidationError, PlottingException) as e:
+                # Expected user error
+                return PlotResult(error=str(e))
+            except alt.MaxRowsError:
+                return PlotResult(
+                    error="The selected dataset is too large to plot. Please try a different plot type or select a fewer number of samples."
+                )
+            except Exception:
+                # TODO
+                # sentry_sdk.capture_exception(e)
+                return PlotResult(
+                    error="Something went wrong. We've been notified about the issue. Please reload the page to try again."
```

Comment:
I know that it is an internal module. But I am not sure if `onecodex` library should contain web app's error messages.